rs/execution_environment/src/scheduler/round_schedule.rs

```diff
@@ -100,6 +100,40 @@ impl RoundSchedule {
         }
     }
 
+    /// Marks idle canisters in front of the schedule as fully executed.
+    pub fn charge_idle_canisters(
+        &self,
+        canisters: &mut BTreeMap<CanisterId, CanisterState>,
+        fully_executed_canister_ids: &mut BTreeSet<CanisterId>,
+        round_id: ExecutionRound,
+        is_first_iteration: bool,
+    ) {
+        for canister_id in self.ordered_new_execution_canister_ids.iter() {
+            let canister = canisters.get_mut(canister_id);
+            if let Some(canister) = canister {
+                let next_execution = canister.next_execution();
+                match next_execution {
+                    NextExecution::None => {
+                        Self::finish_canister_execution(
+                            canister,
+                            fully_executed_canister_ids,
+                            round_id,
+                            is_first_iteration,
+                            0,
+                        );
+                    }
+                    // Skip install code canisters.
+                    NextExecution::ContinueInstallCode => {}
+
+                    NextExecution::StartNew | NextExecution::ContinueLong => {
+                        // Stop searching after the first non-idle canister.
```

Comment:
I would add here that this is a lower-bound approximation of the canisters that would have gotten a "full round" had they had anything to execute. But that's good enough because in the next round the idle canisters we left out will be at the front of the schedule and very likely still idle, so we will charge them at that time.

However, having just written that, I realized there's an issue: if you have a canister with 100 compute allocation that always has something to execute (e.g. a heartbeat), it will always prevent any idle canisters from getting charged until they accumulate more than 100 priority (and the handful of other active canisters accumulate huge negative priority in the process).

Maybe you should stop after the 4th non-idle canister instead? But that's also not accurate, as (due to splitting canisters across cores) you may not even have tried to execute that canister (e.g. it could be the second canister on core 0 and most of the execution may have happened on cores 1, 2 and 3).

Off the top of my head, it seems to me like it makes most sense to keep iterating until the first non-idle canister *after* you've seen all `fully_executed_canister_ids`. Because you could argue that any idle canister that was in-between the active canisters you executed (including the first one you didn't execute for a full round) might have been executed had they had any inputs. I believe this should be a much better estimate, but I'm not 100% sure. I have to think about it for a while.

Reply:
The current approach looks simple and correct. While I agree it has some limitations, it's definitely an improvement for the current situation. Any other complications, such as backtracking or finding the first non-idle canister after all fully_executed_canister_ids, sounds challenging and I'm not sure it's worth implementing now...

Reply:
There's no need to backtrack. It's just a different (and correct) estimate of how far we would have gotten with executing idle canisters.

Your current implementation will stop at the first DTS execution. Or as soon as it encounters a canister that was scheduled first, but didn't execute all its backlog. As described above, it can also stop every single round without ever charging an idle canister (something that we agree is a problem).

My proposed solution is to simply keep going until you've seen all the canisters that got a full execution round. Which makes perfect sense: if you've executed canisters 1, 5 and 7, but not canisters 3 and 9; and all other canisters up to 1000 are idle; then you could argue that canisters 2, 4, 6 and 8 were "executed" they just didn't have any inputs. This is definitely a better guess at the truth than saying "no idle canister was executed" simply because canister 1 did not consume its entire backlog.

Reply:
Maybe I don't understand. Here is an example:

```
CPU1: 1 3 5 7 9 b d
CPU2: 2 4 6 8 a c ...
```

If we fully executed CPU1 (canisters 1, 3, 5, 7, 9) and only canister 2 on CPU2, can we really claim that canisters 4, 6 and 8 were executed? Seems they were simply waiting for the canister 2 to finish on the CPU2...

Resolving the conversation for now...

Reply:
It's not about going to extremes. It's about doing something that makes sense. This particular change makes just as much sense as charging the first 10 idle canisters: sometimes it will match, sometimes not. (Or, as in your example, as charging all idle canisters.)

There is no good reason whatsoever why you should stop at the first non-idle canister.